core/pkg/observability/printer.go

```diff
@@ -2,23 +2,44 @@ package observability
 
 import (
 	"sync"
+	"time"
 )
 
 // Printer stores console messages to display to the user.
 type Printer struct {
 	sync.Mutex
 	messages []string
+
+	// For rate-limited messages, this is the next time a message may be sent.
+	rateLimits map[string]time.Time
```

Comment:
could it be an issue during closing of a stream?

Reply:
can you clarify?

Reply:
yeah, sorry, since it is time delayed, could there be a case where we finish before we read everything. 
Since in the handler get this polling message to read the content of the printer, so it might not be updated and we might miss message?

Reply:
oh, it's not delayed---messages that happen too frequently are completely skipped. Not buffered for later. It's the usual behavior in loggers in other languages with this feature.

Reply:
oh ok, so i misunderstood the code... thanks for the clarification.
yep we use this feature in python for our `term loggers`

Reply:
should've read the tests, lol...